src/streamed/state.ts

```diff
@@ -1,32 +1,443 @@
+}
+
+/**
+ * Amount used to signal that an order is an unlimited order.
+ */
+const UNLIMITED_ORDER_AMOUNT = BigInt(2 ** 128) - BigInt(1)
+
+/**
+ * JSON representation of the account state.
+ */
+export interface AuctionStateJson {
+  tokens: { [key: string]: string },
+  accounts: {
+    [key: string]: {
+      balances: { [key: string]: string },
+      pendingWithdrawals: { [key: string]: { batchId: number, amount: string } },
+      orders: {
+        buyToken: TokenId,
+        sellToken: TokenId,
+        validFrom: number,
+        validUntil: number | null,
+        priceNumerator: string,
+        priceDenominator: string,
+        remainingAmount: string,
+      }[],
+    }
+  },
+}
+
+/**
+ * Manage the exchange's auction state by incrementally applying events.
+ */
+export class AuctionState {
   private lastBlock: number = -1;
 
+  private readonly tokens: Map<TokenId, Address> = new Map();
+  private readonly accounts: Map<Address, Account> = new Map();
+  private lastSolution?: { submitter: Address, burntFees: bigint };
+
   constructor(
     private readonly options: OrderbookOptions,
   ) {}
 
   /**
-   * Apply specified ordered events.
+   * Create an object representation of the current account state for JSON
+   * serialization.
+   */
+  public toJSON(): AuctionStateJson {
+    function map2obj<K extends { toString: () => string }, V, T>(
+      map: Map<K, V>,
+      convert: (value: V) => T,
+    ): { [key: string]: T } {
+      const result: { [key: string]: T } = {}
+      for (const [key, value] of map.entries()) {
+        result[key.toString()] = convert(value)
+      }
+      return result
+    }
+
+    return {
+      tokens: map2obj(this.tokens, addr => addr),
+      accounts: map2obj(this.accounts, account => ({
+        balances: map2obj(account.balances, balance => balance.toString()),
+        pendingWithdrawals: map2obj(
+          account.pendingWithdrawals,
+          withdrawal => ({
+            ...withdrawal,
+            amount: withdrawal.amount.toString(),
+          }),
+        ),
+        orders: account.orders.map(order => ({
+          ...order,
+          priceNumerator: order.priceNumerator.toString(),
+          priceDenominator: order.priceDenominator.toString(),
+          remainingAmount: order.remainingAmount.toString(),
+        })),
+      })),
+    }
+  }
+
+  /**
+   * Retrieves block number the account state is accepting events for.
+   */
+  public get nextBlock(): number {
+    return this.lastBlock + 1
+  }
+
+  /**
+   * Apply all specified ordered events.
    *
    * @remarks
    * This method expects that once events have been applied up until a certain
-   * block, then no new events from that block (or an earlier block) will be
-   * applied.
+   * block number, then no new events for that block number (or an earlier block
+   * number) are applied.
+   *
    */
-  public applyEvents(events: EventData[]): void {
+  public applyEvents(events: EventData[]) {
```

Comment:
> @fleupold Would be nice to pull this into a small method too to keep this switch statement free of logic.

I'd rather submit a follow up PR to clean up the `applyEvents` method so that each case defers to its own method to keep the switch statement "dumb".

Reply:
<a href="https://gitme.me/image?url=https%3A%2F%2Fmedia0.giphy.com%2Fmedia%2F3ohs870THIcgnwwjiU%2Fgiphy-downsized-medium.gif&token=promise" data-gitmeme-token="promise"><img src="https://media0.giphy.com/media/3ohs870THIcgnwwjiU/giphy-downsized-medium.gif" title="Created by gitme.me with /promise"/></a>

Reply:
:rofl: 
My reasoning is this PR is already too big to review so any mistakes I make would probably fly under the radar.

Reply:
This has been addressed in #725